std/experimental/allocator/building_blocks/bitmapped_block.d

```diff
@@ -1746,3 +1760,350 @@ private struct BitVector
+    }
+
+    /**
+    Allocates `s` bytes of memory and returns it, or `null` if memory
+    could not be allocated.
+
+    The `SharedBitmappedBlock` cannot allocate more than the given block size.
+    Allocations are satisfied by searching the first unset bit in the bitmap,
+    and atomically setting it.
+    In rare memory pressure scenarios, the allocation could fail.
+    */
+    @trusted void[] allocate(const size_t s)
+    {
+        import core.atomic : cas, atomicLoad, atomicOp;
+
+        if (s.divideRoundUp(blockSize) != 1)
+            return null;
+
+        // First zero bit position for all values in the 0 - 255 range
+        // for fast lookup
+        static ubyte[] firstZero = [
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2,
+            2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
+            2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
+            4, 4, 4, 4, 4, 4, 4, 4, 5, 5, 5, 5, 6, 6, 7
+        ];
+
+        for (size_t i = 0; i < _control.length; i++)
+        {
+            ulong controlVal, newControlVal, bitIndex;
+            do
+            {
+                bitIndex = 0;
+                newControlVal = 0;
+                controlVal = atomicLoad(_control[i]);
+
+                // skip all control words which have all bits set
+                if (controlVal == ulong.max)
+                    break;
+
+                // fast lookup of first byte which has at least one zero bit
+                foreach (byteIndex; 0 .. 8)
+                {
+                    ulong mask = (0xFFUL << (8 * (7 - byteIndex)));
+                    if ((mask & controlVal) != mask)
+                    {
+                        ubyte byteVal = cast(ubyte) ((mask & controlVal) >> (8 * (7 - byteIndex)));
+                        bitIndex += firstZero[byteVal];
+                        newControlVal = controlVal | (1UL << (63 - bitIndex));
+                        break;
+                    }
+                    bitIndex += 8;
+                }
+            } while (!cas(&_control[i], controlVal, newControlVal));
+
+            auto blockIndex = bitIndex + 64 * i;
+            if (controlVal != ulong.max && blockIndex < _blocks)
+            {
+                size_t payloadBlockStart = cast(size_t) blockIndex * blockSize;
```

Comment:
This could overflow. How about using `std.experimental.checkedint`?

Reply:
An extra check here could be expensive.
Also the blockIndex check above itself prevents against overflow